nifi-external/nifi-storm-spout/src/main/java/org/apache/nifi/storm/NiFiBolt.java

```diff
@@ -0,0 +1,190 @@
+
+        LOGGER.info("Bolt is prepared with Batch Size " + batchSize
+                + ", Batch Interval " + batchIntervalInSec
+                + ", Tick Frequency is " + tickFrequencySeconds);
+    }
+
+    protected SiteToSiteClient createSiteToSiteClient() {
+        return new SiteToSiteClient.Builder().fromConfig(clientConfig).build();
+    }
+
+    @Override
+    public void execute(Tuple tuple) {
+        if (TupleUtils.isTick(tuple)) {
+            // if we have a tick tuple then lets see if enough time has passed since our last batch was processed
+            if ((System.currentTimeMillis() / 1000 - lastBatchProcessTimeSeconds) >= batchIntervalInSec) {
+                LOGGER.debug("Received tick tuple and reached batch interval, executing batch");
+                finishBatch();
+            } else {
+                LOGGER.debug("Received tick tuple, but haven't reached batch interval, nothing to do");
+            }
+        } else {
+            // for a regular tuple we add it to the queue and then see if our queue size exceeds batch size
+            this.queue.add(tuple);
+
+            int queueSize = this.queue.size();
+            LOGGER.debug("Current queue size is " + queueSize + ", and batch size is " + batchSize);
+
+            if (queueSize >= batchSize) {
+                LOGGER.debug("Queue Size is greater than or equal to batch size, executing batch");
+                finishBatch();
+            }
+        }
+    }
+
+    private void finishBatch() {
+        LOGGER.debug("Finishing batch of size " + queue.size());
+        lastBatchProcessTimeSeconds = System.currentTimeMillis() / 1000;
+
+        final List<Tuple> tuples = new ArrayList<>();
+        queue.drainTo(tuples);
+
+        if (tuples.size() == 0) {
+            LOGGER.debug("Finishing batch, but no tuples so returning...");
+            return;
+        }
+
+        try {
+            final Transaction transaction = client.createTransaction(TransferDirection.SEND);
+            if (transaction == null) {
+                throw new IllegalStateException("Unable to create a NiFi Transaction to send data");
+            }
+
+            // convert each tuple to a NiFiDataPacket and send it as part of the transaction
+            for (Tuple tuple : tuples) {
+                final NiFiDataPacket dataPacket = builder.createNiFiDataPacket(tuple);
+                transaction.send(dataPacket.getContent(), dataPacket.getAttributes());
+            }
+
+            transaction.confirm();
+            transaction.complete();
+
+            // ack the tuples after successfully completing the transaction
+            for (Tuple tuple : tuples) {
+                collector.ack(tuple);
+            }
+
+        } catch(Exception e){
+            LOGGER.warn("Unable to process tuples due to: " + e.getMessage(), e);
+            for (Tuple tuple : tuples) {
+                collector.fail(tuple);
```

Comment:
I suppose it could, but I don't think there is anything you could do if it did fail. From looking at bolts provided in the Storm codebase they don't usually wrap the ack/fail in a try/catch. I think if an ack/fail is not sent for a given tuple after a certain amount of time then the framework tries to resend that tuple.